ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -501,6 +503,38 @@ static int HandleVnodeOperation(
                     goto CleanupAndReturn;
                 }
             }
+            
+            if (ActionBitIsSet(action, KAUTH_VNODE_WRITE_DATA))
+            {
+                if (!TryGetVirtualizationRoot(&perfTracer, context, currentVnode, pid, &root, &vnodeFsidInode, &kauthResult, kauthError))
+                {
+                    goto CleanupAndReturn;
+                }
+                
+                PrjFSFileXAttrData rootXattr = {};
+                SizeOrError xattrResult = Vnode_ReadXattr(currentVnode, PrjFSFileXAttrName, &rootXattr, sizeof(rootXattr));
+                if (xattrResult.error == ENOATTR)
```

Comment:
For my own understanding, when does the extended attribute get removed from the file?

Reply:
It's [removed when the provider receives a modified file notification](https://github.com/Microsoft/VFSForGit/blob/ed362aec65f790f26a2a1f1c593f7602687388f3/ProjFS.Mac/PrjFSLib/PrjFSLib.cpp#L992).

Reply:
@kewillford I'm really glad you brought this up, and I think it would good to make sure everyone is aware of how the process works (and confirm my own understanding).

When a new placeholder is written we set FileFlags_IsEmpty=true and set PrjFSFileXAttrName.  
When the placeholder becomes hydrated, FileFlag_isEmpty=false
When the hydrated file is modified we remove PrsFSFileXAttrName

@wilbaker is that correct?

PreConvertToFull means we should add the file to modifiedFiles.dat.  A Modified alert tells us to run the statusCache again (file would already be in modifiedFiles.dat).  If my understanding is correct the name PreCovertToFull seems a bit off, as does the partialFile variable name on  PrsFSFileXAttrName.

Reply:
This matches my (possibly incorrect) understanding, and I too am confused by the apparent "placeholder" vs "partial" terminology mismatch. Is there a difference?

Reply:
> When a new placeholder is written we set FileFlags_IsEmpty=true and set PrjFSFileXAttrName.
When the placeholder becomes hydrated, FileFlag_isEmpty=false
When the hydrated file is modified we remove PrsFSFileXAttrName

Yep that's correct.

> PreConvertToFull means we should add the file to modifiedFiles.dat.

That's right too.  When a file is converted to full (i.e. a plain\regular file without our flags or xattrs) then git becomes responsible for the file and it must be added to ModifiedPaths.dat.

It's also important to note that placeholders are removed from the placeholder list in response to this callback.

> If my understanding is correct the name PreConvertToFull seems a bit off

Which part of the name seems off?  This callback lets the provider know the placeholder is about to be converted to full.

In addition to the topic of naming, this comment brings up a great point.  I missed a detail of PreConvertToFull's behavior on Windows in our offline conversations. 

On Windows the flow is like this:

1. Placeholder file is being opened for write
2. Provider's PreConvertToFull callback is called, and the provider either returns HResult.OK, or an error code
3. If the provider returns OK, then the placeholder reparse data is removed from the file (and it's converted to full)
4. If the provider returns an error, the file is *not* converted to full, and the caller (trying to write to the file) sees an error

So on Windows the provider has a chance to deny the placeholder from being converted to full.  Although VFS4G does have a window of time where it does return an error, we have #793 open for eliminating that window (and so I'm not sure we need\want that feature in the kext).

> as does the partialFile variable name on PrsFSFileXAttrName.
> This matches my (possibly incorrect) understanding, and I too am confused by the apparent "placeholder" vs "partial" terminology mismatch. Is there a difference?

*Unofficial Glossary*

- "Placeholder" -> A file that is tagged with ProjFS attributes (reparse data on Windows and xattrs on Mac).  
- "Empty Placeholder" -> A placeholder that does not have its file contents
- "Hydrated Placeholder" or "Partial File" -> A placeholder whose file contents are on disk.  The Windows ProjFS API is designed such that in the future a provider could "partially" hydrate a file and only provide the parts of the file that the caller cares about, but VFS4G has no plans to use this feature.
- "Full file" -> A regular filesystem file that does not have any ProjFS related attributes attached to it.  The term "full file" is used to describe both placeholders that were converted to full, and newly created files (that were never projected by VFS4G).

I have no problem with eliminating the term "partial file" from the codebase if that makes the code easier to follow.

Reply:
> Which part of the name seems off? This callback lets the provider know the placeholder is about to be converted to full.

I was thinking PreConvertToFull was in relation to an empty placeholder being hydrated.  Now I understand the meaning is non-ritualized 'full' file.  Thanks for the glossary it all makes sense now!